hbase-server/src/main/java/org/apache/hadoop/hbase/master/normalizer/AbstractRegionNormalizer.java

```diff
@@ -170,6 +170,8 @@ private static boolean skipForMerge(final RegionState state) {
         + "number of regions: {}",
       table, avgRegionSize, table, tableRegions.size());
 
+    // Sorting based on RegionInfo.COMPARATOR.
+    tableRegions.sort(RegionInfo.COMPARATOR);
```

Comment:
It does maintain a sort list, which is based on regionName (not like in branch-1, it is based on RegionInfo.COMPARATOR). With regionName, it will have out of order for STARTKEY. 'aa1' and 'aa1!' will have order of ('aa1!', 'aa1'), as 'aa1''s next byte is delimiter ','. The only place I checked for ordering based on STARTKEY is normalizer.

Reply:
Looking around branch-2.3, I think the normalizer is the only place where the region order has some assumed meaning.

Reply:
Yep, it exists in branch-2.2 as well.